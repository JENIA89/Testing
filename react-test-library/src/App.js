import { useEffect, useState } from "react";


function App() {
    const [data, setData] = useState(null)
    const [toggle, setToggle] = useState(false)

    const onClick = () => setToggle(prev => !prev)

    useEffect(() => {
      const id = setTimeout(() => {
        setData({})
      }, 1000);
    
      return () => {
        clearTimeout(id)
      }
    }, [data])
    
  return (
    <div>
        {toggle &&  <div data-testid='toggle-elem'>toggle</div>}
        {data && <div style={{color: 'red'}}>data</div>}
        <h1>Hello world</h1>
        <button data-testid='toggle-btn' onClick={onClick}>click me</button>
        <input type='text' placeholder='input value'/>
    </div>
  );
}

export default App;
