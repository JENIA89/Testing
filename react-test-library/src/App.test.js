import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from './App';

describe('TEST APP', () => {
    test('get element test', () => {
        render(<App />);
        const helloWorldElem = screen.getByText(/hello world/i);
        const btn = screen.getByRole('button');
        const input = screen.getByPlaceholderText(/input value/i);
        expect(helloWorldElem).toBeInTheDocument();
        expect(btn).toBeInTheDocument();
        expect(input).toBeInTheDocument();
        expect(input).toMatchSnapshot();
      });

      test('async test', async () => {
        render(<App />);
        // query используется в основном для подтверждения отсутствия элемента
        // const helloWorldElem = screen.queryByText(/hello2/i);
        // expect(helloWorldElem).toBeNull()
        const helloWorldElem = await screen.findByText(/data/i);
        expect(helloWorldElem).toBeInTheDocument();
        expect(helloWorldElem).toHaveStyle({color: 'red'});
      });

      test('onclick test', async () => {
        render(<App />);
        const btn = screen.getByTestId('toggle-btn');
        expect(screen.queryByTestId('toggle-elem')).toBeNull();
        fireEvent.click(btn);
        expect(screen.queryByTestId('toggle-elem')).toBeInTheDocument();
        fireEvent.click(btn);
        expect(screen.queryByTestId('toggle-elem')).toBeNull();
      });

      test('input test', async () => {
        render(<App />);
        const input = screen.getByPlaceholderText(/input value/i);
        expect(screen.queryByTestId('value-elem')).toContainHTML('');
        // fireEvent искуственное событие
        // fireEvent.input(input, {
        //     target: { value:'123123' }
        // })
        // userEvent работает с событиями пользователя
        userEvent.type(input, '123123')
        expect(screen.queryByTestId('value-elem')).toContainHTML('123123');
      });
})
