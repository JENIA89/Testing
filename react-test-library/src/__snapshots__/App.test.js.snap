// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST APP get element test 1`] = `
<input
  placeholder="input value"
  type="text"
/>
`;

exports[`TEST APP renders learn react link 1`] = `
<input
  placeholder="input value"
  type="text"
/>
`;

exports[`TEST APP renders react ui 1`] = `
<input
  placeholder="input value"
  type="text"
/>
`;

exports[`renders learn react link 1`] = `
<input
  placeholder="input value"
  type="text"
/>
`;
